### PR TITLE
Vite/Addon-docs: Fix customization of MDX plugins

### PR DESCRIPTION
**Pull request by joshwooding**
~Blocked by https://github.com/storybookjs/mdx2-csf/pull/28~
Replaces https://github.com/storybookjs/storybook/pull/20096
Closes https://github.com/storybookjs/storybook/pull/20116

Issue:

## How to test

- [ ] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [ ] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

File: code/addons/docs/src/preset.ts [hunk 1]
```diff
@@ -61,15 +63,22 @@ async function webpack(
...
   } = options;
 
-  const mdxLoaderOptions = await options.presets.apply('mdxLoaderOptions', {
+  const mdxLoaderOptions: CompileOptions = await options.presets.apply('mdxLoaderOptions', {
     skipCsf: true,
+    ...mdxPluginOptions,
     mdxCompileOptions: {
       providerImportSource: '@storybook/addon-docs/mdx-react-shim',
-      remarkPlugins: [remarkSlug, remarkExternalLinks],
+      // @ts-expect-error CompileOptions type is wrong
```
Comment (joshwooding) [inline code/addons/docs/src/preset.ts:76]: https://github.com/storybookjs/mdx2-csf/issues/29

File: code/addons/docs/src/preset.ts [hunk 2]
```diff
@@ -61,15 +63,22 @@ async function webpack(
...
     csfPluginOptions = {},
+    mdxPluginOptions = {},
     sourceLoaderOptions = null,
     transcludeMarkdown = false,
   } = options;
 
-  const mdxLoaderOptions = await options.presets.apply('mdxLoaderOptions', {
+  const mdxLoaderOptions: CompileOptions = await options.presets.apply('mdxLoaderOptions', {
     skipCsf: true,
+    ...mdxPluginOptions,
```
Comment (joshwooding) [inline code/addons/docs/src/preset.ts:73]: @shilman How do you think we should merge these configs?

This setup always respects skipCsf but merges everything else

File: code/lib/builder-vite/src/plugins/mdx-plugin.ts [hunk 3]
```diff
@@ -46,19 +48,35 @@ export function mdxPlugin(options: Options): Plugin {
...
+
+      const { mdxPluginOptions } = docsOptions;
 
       const mdxLoaderOptions = await options.presets.apply('mdxLoaderOptions', {
+        skipCsf: !storiesFile,
+        ...mdxPluginOptions,
         mdxCompileOptions: {
           providerImportSource: '@storybook/addon-docs/mdx-react-shim',
+          ...mdxPluginOptions.mdxCompileOptions,
+          remarkPlugins: [remarkSlug, remarkExternalLinks].concat(
```
Comment (joshwooding) [inline code/lib/builder-vite/src/plugins/mdx-plugin.ts:66]: Should we move these plugins to mdx2-csf? 

Seems to be a couple issues asking for the removal of it https://github.com/storybookjs/storybook/issues/9329

**Comment by IanVS** [comment]
I see this is still a draft, what is left to do before it is ready for review?

**Comment by joshwooding** [comment]
> I see this is still a draft, what is left to do before it is ready for review?

Was hoping to get some guidance on my earlier questions

**Review by shilman — COMMENTED** [review]
@joshwooding This is the current Webpack configuration in 7.0

https://github.com/storybookjs/storybook/blob/next/code/addons/docs/src/preset.ts#L46-L53

I'm wondering how this change squares with that. It's also possible that we made an unintentional breaking change there & we need to restore some level of configuration there that we deleted with the last round of changes. WDYT?

**Comment by ndelangen** [comment]
@shilman let's discuss tomorrow?

File: code/lib/builder-vite/src/plugins/mdx-plugin.ts [hunk 4]
```diff
@@ -15,10 +15,8 @@ const isStorybookMdx = (id: string) => id.endsWith('stories.mdx') || id.endsWith
 export async function mdxPlugin(options: Options): Promise<Plugin> {
   const include = /\.mdx?$/;
   const filter = createFilter(include);
-  const addons = await options.presets.apply<StorybookConfig['addons']>('addons', []);
```
Comment (joshwooding) [inline code/lib/builder-vite/src/plugins/mdx-plugin.ts:18]: @shilman addons wasn't defined when doing this. When I debugged the code it seems like I can get the options directly?

**Comment by joshwooding** [comment]
I've tested this locally with:

```
 { name: "@storybook/addon-docs",
      options: {
      mdxPluginOptions: {
        mdxCompileOptions: {
          remarkPlugins: [remarkGfm]
        }
        }
      }
    }
```
and it works

**Review by ndelangen — APPROVED** [review]

File: code/lib/builder-vite/src/plugins/mdx-plugin.ts [hunk 5]
```diff
@@ -15,10 +15,10 @@ const isStorybookMdx = (id: string) => id.endsWith('stories.mdx') || id.endsWith
...
-  const docsOptions =
-    // @ts-expect-error - not sure what type to use here
-    addons.find((a) => [a, a.name].includes('@storybook/addon-docs'))?.options ?? {};
+  const { mdxPluginOptions, jsxOptions } = await options.presets.apply<Record<string, any>>(
```
[the remark this answers is not in the document]
Comment (joshwooding) [reply code/lib/builder-vite/src/plugins/mdx-plugin.ts:18]: Yeah, I was quite surprised too. I only found it when debugging through the code.

**Comment by ndelangen** [comment]
@joshwooding you're a hero! ❤️

**Pull request merged**